Remove alternate file stream code from non-Windows

SteveL-MSFT opened the pull request:
AlternateStreams support relies on pinvoke to win32 apis which don't work

on non-Windows and produces errors about not loading a dll in the case of copy-item

Fix https://github.com/PowerShell/PowerShell/issues/4542

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to CONTRIBUTING.md.

-->

SteveL-MSFT committed:
AlternateStreams support relies on pinvoke to win32 apis which don't work
on non-Windows and produces errors about not loading a dll in the case
of copy-item

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/UnblockFile.cs:
```diff
@@ -13,6 +13,7 @@
 
 #endregion
 
+#if !UNIX
```
`Unblock-File` is excluded from Unix at all so we don't need to exclude the code here.

SteveL-MSFT replied:
That file is still being compiled on Unix

iSazonov replied:
I think we should exclude them in csproj.

SteveL-MSFT replied:
I'm fine with that but then we have effectively conditional compilation in both the source and the csproj files

dantraMSFT replied:
Should we rename this to UnblockFile.Windows.cs? 

---
In reply to: [133130873](https://github.com/PowerShell/PowerShell/pull/4567#discussion_r133130873) [](ancestors = 133130873)

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1:
```diff
@@ -126,4 +126,17 @@ Describe "Get-Content" -Tags "CI" {
+
+  It "Should support colons in filename on Linux/Mac" -Skip:($IsWindows) {
+    Set-Content "${testPath}:Stream" -Value "Hello"
+    Get-Content "${testPath}:Stream" | Should BeExactly "Hello"
```
Please add a check that `${testPath}:Stream` is a file name.

SteveL-MSFT replied:
Doesn't `Get-Content` implicitly cover that?

SteveL-MSFT replied:
I guess if the Content cmdlets were doing something funny it could mask an issue, I'll add a check

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1:
```diff
@@ -126,4 +126,17 @@ Describe "Get-Content" -Tags "CI" {
+    Set-Content "${testPath}:Stream" -Value "Hello"
+    Get-Content "${testPath}:Stream" | Should BeExactly "Hello"
+  }
+
```
Extra line.

SteveL-MSFT replied:
Will remove

iSazonov reviewed (commented):
Leave a comment

adityapatwardhan commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -6828,10 +6843,11 @@ public void ClearContent(string path)
             {
+#if !UNIX
                 bool clearStream = false;
+                string streamName = null;
```
Unnecessary change of line.

SteveL-MSFT replied:
The change of grouping of the variables makes it a bit more readable to me

adityapatwardhan replied:
Closed.

adityapatwardhan commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -147,20 +147,15 @@ internal static SecurityZone GetFileSecurityZone(string filePath)
             Diagnostics.Assert(Path.IsPathRooted(filePath), "Caller makes sure the path is rooted.");
             Diagnostics.Assert(Utils.NativeFileExists(filePath), "Caller makes sure the file exists.");
-#if CORECLR
             string sysRoot = System.Environment.GetEnvironmentVariable("SystemRoot");
```
Should this block be surrounded by #if !UNIX?

SteveL-MSFT replied:
Some of this code is unnecessary on UNIX so I'll wrap it

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1:
```diff
@@ -126,4 +126,17 @@ Describe "Get-Content" -Tags "CI" {
+
+  It "Should support NTFS streams using colon syntax" -Skip:(!$IsWindows) {
+    Set-Content "${testPath}:Stream" -Value "Foo"
+    { Test_path "${testPath}:Stream" | ShouldBeErrorId "ItemExistsNotSupportedError,Microsoft.PowerShell.Commands,TestPathCommand" }
```
Should this be ```Test-Path``` instead of ```Test_path```

SteveL-MSFT replied:
Yes, odd that this test passed

SteveL-MSFT committed:
address PR feedback

adityapatwardhan commented on src/Microsoft.PowerShell.Commands.Utility/Microsoft.PowerShell.Commands.Utility.csproj:
```diff
@@ -35,6 +35,7 @@
     <Compile Remove="commands\utility\ShowCommand\ShowCommandParameterType.cs" />
     <Compile Remove="commands\utility\ShowCommand\ShowCommandProxy.cs" />
     <Compile Remove="commands\utility\update-list.cs" />
+    <Compile Remove="commands\utility\Unblockfile.cs" Condition=" '$(Configuration)' == 'Linux' " />
```
Should we add ```#if !UNIX``` to the start of the file like: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/utils/tracing/EtwActivity.cs#L1

That seems more discoverable to me when doing code reviews or debugging.

SteveL-MSFT replied:
That's what I had originally.  Did a search and I don't see other cases of conditional compilation, so I think it's more consistent to use a whole file `#ifdef`.  @iSazonov ok?

iSazonov replied:
I think it's technically better to exclude files in csproj files. But currently we don't use a lot of opportunities of MSBuild and also we are cleaning up FullCLR files and respectively remove "Compile Remove" from csproj files. So it is Ok to use  #ifdef .

/cc @daxian-dbw  Could you comment too?

daxian-dbw replied:
I don't really have a preference although `csproj` approach seems more discoverable. However in this case, I don't think `Unblockfile.cs` should be completely excluded from Unix build as it will increase the number of `#if unix/endif` blocks in places where the code from this file is used, which will impact readability.

dantraMSFT commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1146,10 +1147,12 @@ protected override void GetItem(string path)
                 FileSystemInfo result = GetFileSystemItem(path, ref isContainer, false);
                 if (result != null)
                 {
+#if !UNIX
```
This #if/#endif block is cumbersome to read.  Suggest instead:
if (result != null)
{
#if UNIX
     WriteItemObject(...)
#else
    original code
 }

SteveL-MSFT replied:
The `WriteItemObject` code applies to both UNIX and !UNIX.  If I move that bit of code up, I'd need more `#ifdefs` making it less readable

dantraMSFT replied:
Actually, if you don't mind duplicating the WriteItemObject line, it can be a pure #if #else block.  Your call.

SteveL-MSFT replied:
I think the other case I was fine with a bit of duplication, but prefer to leave this one as-is.

dantraMSFT commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -2798,12 +2804,17 @@ protected override void RemoveItem(string path, bool recurse)
                     return;
                 }
 
+#if !UNIX
```
As above, the code is easier to read if you use a #if UNIX and contain the UNIX lines completely.  Its a simple if/else with a single statement for UNIX but  looks unwieldy with the current #if/#else placement.

SteveL-MSFT replied:
I can move some code around so the UNIX code is in one place and the !UNIX code is also together.

dantraMSFT commented:
>                     if (isAlternateDataStream)

Should this also be conditioned on !UNIX?

---
Refers to: src/System.Management.Automation/namespaces/FileSystemProvider.cs:4270 in 7209973. [](commit_id = 720997351e1de1cf26593cfe0e95699d8b3b9eb3, deletion_comment = False)

dantraMSFT commented:
>             return Platform.NonWindowsIsSameFileSystemItem(pathOne, pathTwo);

Shouldn't there be a Platform.IsSameFileSystemItem?

---
Refers to: src/System.Management.Automation/namespaces/FileSystemProvider.cs:8274 in 7209973. [](commit_id = 720997351e1de1cf26593cfe0e95699d8b3b9eb3, deletion_comment = False)

dantraMSFT commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -147,20 +147,17 @@ internal static SecurityZone GetFileSecurityZone(string filePath)
-#else
-            return MapSecurityZoneWithUrlmon(filePath);
 #endif
+            return MapSecurityZoneWithoutUrlmon(filePath);
```
it appears you're still hitting alternate data streams on unix.
ClrFacade:  MapSecurityZoneWithoutUrlmon -> ReadFromZoneIdentifierDataStream

dantraMSFT commented on src/System.Management.Automation/namespaces/FileSystemContentStream.cs:
```diff
@@ -843,7 +843,9 @@ private void CreateStreams(string filePath, string streamName, FileMode fileMode
             {
                 if (!String.IsNullOrEmpty(streamName))
                 {
+#if !UNIX
```
On UNIX, _stream will be null causing a ArgumentNullException from new StreamReader at line 874.  At best, it will be misleading. Consider throwing NotSupportException on UNIX in the associated constructors

SteveL-MSFT replied:
makes sense, will change

SteveL-MSFT committed:
address PR feedback

SteveL-MSFT commented:
@dantraMSFT I addressed your feedback.  Regarding `Platform.IsSameFileSystem` we should probably have a single API that is platform agnostic, but that's out of scope of this PR.

dantraMSFT commented on src/System.Management.Automation/namespaces/FileSystemContentStream.cs:
```diff
@@ -845,6 +845,8 @@ private void CreateStreams(string filePath, string streamName, FileMode fileMode
 #if !UNIX
                     _stream = AlternateDataStreamUtilities.CreateFileStream(filePath, streamName, fileMode, fileAccess, fileShare);
+#else
+                    throw new PlatformNotSupportedException();
```
I think these exceptions need to be explicit in indicating alternate data streams are not supported. That's why I suggested an exception at ctor time; an exception that indicates streamName is not supported on UNIX.

SteveL-MSFT replied:
see @daxian-dbw comment below as better way to address this

daxian-dbw commented on src/System.Management.Automation/utils/ClrFacade.cs:
```diff
@@ -147,20 +147,17 @@ internal static SecurityZone GetFileSecurityZone(string filePath)
         {
```
SecurityZone is a windows concept which I think shouldn't be applied when running on Unix plats. We shouldn't have `#if UNIX/endif` in this method or the methods exclusively used by this method, instead, we probably should analyze the call sites of this method to make sure this method doesn't run on unix plats, or better, enclose this method within `#if !UNIX/endif`.

SteveL-MSFT replied:
There are some other checks this code does, however, upon closer inspection, I think they also don't apply to Unix so I'll `#ifdef` this code out on UNIX

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1117,6 +1117,7 @@ protected override void GetItem(string path)
             {
+#if !UNIX
                 bool retrieveStreams = false;
                 FileSystemProviderGetItemDynamicParameters dynamicParameters = null;
```
I think the fix should be to make the dynamic parameters related to `Stream` operation unavailable in FileSystemProvider. Otherwise, it's broken user experience as I can use the `-Stream` parameter but it doesn't behave as expected.

JamesWTruher replied:
yes, it would be better to not have the `-stream` parameter if it's not supported. It may also be cleaner to do it this way

daxian-dbw replied:
Once we stop providing the `-Stream` dynamic parameter for FileSystemProvider on Unix, we don't need to add so many `#if UNIX/endif` blocks in this file because those codes will not be hit. I think that will make the code more readable and easy to maintain.

SteveL-MSFT replied:
So we still want the code compiled on UNIX?

daxian-dbw replied:
If that makes the code cleaner overall, then I'm fine to have the code compiled on Unix.

SteveL-MSFT replied:
Looking at the existing code, the FileSystemProvider has a ton of `#if UNIX` and `#if !UNIX` already because the filesystems are inherently different (beyond just NTFS streams).  I think refactoring this code is an exercise best left for the future.

daxian-dbw replied:
I thought about it more and now have this question: which the following experience is better -- copy `Get-Content -Path <file> -Stream <stream>` to Linux powershell and see the error `"A parameter cannot be found"`, or `"Alternate stream is not supported on this platform"`? I think the latter message is better. To achieve this, maybe we can make the setter of the `Stream` property throw `PlatformNotSupportedException` on non-windows.

As for `#if UNIX/endif` blocks, IMHO, we don't necessarily need to have all codes that only apply to a specific platform guarded by the compile definition. I think it's OK to keep such codes around even for other platforms, and the guideline should be readability and maintainability. I still feel it's better to avoid most of the `#if UNIX` blocks by controlling the entry points:
1. Make sure `-Stream` not available on Unix platforms
2. Throw error when the inline stream syntax is used for `-Path` on Unix platforms
3. Skip the alternate stream discovery codes during remote copy (both `PerformCopyFileFromRemoteSession` and `PerformCopyFileToRemoteSession`) on Unix platforms.

After controlling those entry points (and put some comments for sure), it's safe to know that the alternate stream related code paths will not be hit on Unix platforms.

SteveL-MSFT replied:
With the current changes, you don't get any error using `copy-item` going from Linux->Windows nor Windows->Linux

I created a new issue https://github.com/PowerShell/PowerShell/issues/4633 to clean up the code later since filesystemprovider.cs already has 21 `#if` statements before I made any changes

lzybkr replied:
Which is more likely - somebody copying a script from Windows to Linux that uses `-Stream`, or somebody on Linux using parameter completion, sees `-Stream` as a possible option, and wastes time trying to learn what it does?

I'm thinking it's better overall to not suggest parameters on platform if they aren't actually supported.

SteveL-MSFT replied:
`Copy-Item` doesn't expose a `-Stream` param at all.  In the case of copying a NTFS file that contains alternate streams, those streams don't exist on the Linux destination.

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemContentStream.cs:
```diff
@@ -843,7 +843,11 @@ private void CreateStreams(string filePath, string streamName, FileMode fileMode
+#if !UNIX
                     _stream = AlternateDataStreamUtilities.CreateFileStream(filePath, streamName, fileMode, fileAccess, fileShare);
+#else
+                    throw new PlatformNotSupportedException();
```
I think it's better to make sure `CreateStreams` is never used for an alternate stream instead of inserting PlatformNotSupportedException.

SteveL-MSFT replied:
Good idea, will change.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Get-Content.Tests.ps1:
```diff
@@ -126,4 +126,18 @@ Describe "Get-Content" -Tags "CI" {
     } else {$expected = "Hell","o,ll","ll","Worll","d`nHell","o2,ll","ll","Worll","d2`n"}
     for ($i = 0; $i -lt $result.Length ; $i++) { $result[$i]  | Should BeExactly $expected[$i]}
   }
+
```
if you approach this by removing the dynamic parameter on ! windows, we should probably have an explicit test for the parameter - windows => parameter present ; ! windows => parameter absent

SteveL-MSFT replied:
will add

SteveL-MSFT replied:
Need to check if there are already tests using -stream otherwise I'll add

PaulHigin commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -4667,6 +4688,7 @@ private bool PerformCopyFileToRemoteSession(FileInfo file, string destinationPat
 
             bool targetSupportsAlternateStreams = RemoteTargetSupportsAlternateStreams(ps, remoteFilePath);
 
+#if !UNIX
```
The PSCopyToSessionHelper script and PSCopyFromRemoteSession script for remote file copy contains support for alternate streams.  Ideally we would have versions of this script for Linux platforms that remove that support.  But I am not sure it should be part of this change.

SteveL-MSFT replied:
I think that code cleanup should be part of a different PR.  Created https://github.com/PowerShell/PowerShell/issues/4601

PaulHigin commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -8260,6 +8284,7 @@ internal static bool IsSameFileSystemItem(string pathOne, string pathTwo)
         }
 
+#if !UNIX
         internal static bool WinIsSameFileSystemItem(string pathOne, string pathTwo)
```
This method can be made private access since it is only used here.

SteveL-MSFT replied:
will do

PaulHigin commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -8297,6 +8323,7 @@ internal static bool GetInodeData(string path, out System.ValueTuple<UInt64, UIn
         }
 
+#if !UNIX
         internal static bool WinGetInodeData(string path, out System.ValueTuple<UInt64, UInt64> inodeData)
```
Likewise, should be private access.

SteveL-MSFT replied:
will do

SteveL-MSFT commented on src/System.Management.Automation/namespaces/FileSystemContentStream.cs:
```diff
@@ -854,7 +858,11 @@ private void CreateStreams(string filePath, string streamName, FileMode fileMode
+#if !UNIX
                     _stream = AlternateDataStreamUtilities.CreateFileStream(filePath, streamName, fileMode, requestedAccess, fileShare);
+#else
+                    throw new PlatformNotSupportedException();
```
ditto

SteveL-MSFT commented on src/System.Management.Automation/engine/Modules/ImportModuleCommand.cs:
```diff
@@ -1304,7 +1304,9 @@ private static bool IsCmdletizationFile(RemoteDiscoveryHelper.CimModuleFile cimM
                     fullPath,
                     file.RawFileData);
 
+#if !UNIX
```
I believe CDXML cmdlets can work on Linux via OMI

SteveL-MSFT committed:
address PR feedback, remove -Stream parameter for Unix
added new -stream tests

SteveL-MSFT commented:
I believe all feedback has been addressed

PaulHigin reviewed (approved):
LGTM

dantraMSFT commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -4667,6 +4688,7 @@ private bool PerformCopyFileToRemoteSession(FileInfo file, string destinationPat
 
             bool targetSupportsAlternateStreams = RemoteTargetSupportsAlternateStreams(ps, remoteFilePath);
```
Shouldn't this be in the #if !UNIX block?

SteveL-MSFT replied:
yes, will fix

SteveL-MSFT replied:
will fix

SteveL-MSFT replied:
will fix

daxian-dbw commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -1146,10 +1147,12 @@ protected override void GetItem(string path)
                 {
+#if !UNIX
                     // If we want to retrieve the file streams, retrieve them.
                     if (retrieveStreams)
```
The alternate streams can be specified using `-Stream` or an inline stream syntax `"file::stream"`, the check of the inline stream syntax is done in 7 places in `FileSystemProvider.cs` but some of them are not within `#if !UNIX` block.

daxian-dbw replied:
And when running `Get-Content -Path file::stream` on Linux, it probably should throw an error message about "alternate stream is not supported" instead of working unexpectedly or fail with other errors.

SteveL-MSFT replied:
On Unix, a colon is completely allowed in a filename, so we want that to continue working as just a filename.  Tests were added to make sure this works.

I'll fix other cases of using the `file:stream` syntax.

Commit:
ifdef out alternate stream syntax checks not applicable to unix

SteveL-MSFT commented:
@TravisEz13 fixed

SteveL-MSFT commented:
@TravisEz13 @daxian-dbw feedback addressed
